### PR TITLE
Object detection

### PR DESCRIPTION
**Pull request by OvsiienkoVladyslava**
Add and test 2 object detection models

### CODE REVIEW

File: pytorch_object_detection/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# computer_vision
+Test and try CV models
```
Comment (AlexShashko) [inline pytorch_object_detection/README.md]: Some information about experiments? Links to tutorials? Information about author?

File: pytorch_object_detection/detection_models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+from torchvision.utils import draw_bounding_boxes
+from torchvision.transforms.functional import to_pil_image
+
+
+class DetectionPipeline:
+    model: torchvision.models.detection
+    weights: WeightsEnum
+    preprocess_pipeline: torchvision.transforms._presets.ObjectDetection
+    model_classes: list
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Move to __init__(...)

File: pytorch_object_detection/detection_models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import torchvision
+import torch
+from torchvision.io.image import read_image
+from torchvision.models import WeightsEnum
+
+from torchvision.utils import draw_bounding_boxes
+from torchvision.transforms.functional import to_pil_image
+
+
+class DetectionPipeline:
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Is it base class? Should it be abstract?

File: pytorch_object_detection/detection_models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    def __init__(self):
+        self.model.eval()
+        self.model_classes = self.weights.meta["categories"]
+        self.preprocess_pipeline = self.weights.transforms()
+
+    def get_model_info(self):
+        pass
+
+    def _image_preprocess(self, image_paths: list[str] | str):
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Use typing lib
```suggestion
    def _image_preprocess(self, image_paths: List[str] | str):
```

File: pytorch_object_detection/detection_models/__init__.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                    labels.append(self.model_classes[output["labels"][ind]])
+                    boxes.append(output['boxes'][ind])
+            output_scores.append(scores)
+            output_labels.append(labels)
+            output_boxes.append(boxes)
+
+        # Visualize detection results
+        self.visualize_result(raw_images, output_boxes, output_labels)
+
+        return output_labels, output_boxes, output_scores
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Add new line

File: pytorch_object_detection/detection_models/__init__.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    def visualize_result(imgs, boxes, labels):
+        for ind, img in enumerate(imgs):
+            box = torch.stack(boxes[ind], dim=0)
+            drawed_boxes = draw_bounding_boxes(img, boxes=box,
+                                               labels=labels[ind],
+                                               width=2)
+            im = to_pil_image(drawed_boxes.detach())
+            im.show(title=f'Image №{ind}')
+
+    def run(self, imagepaths: list[str] | str, score_threshold=0.9):
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Use the proper typing style

File: pytorch_object_detection/detection_models/configs.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from torchvision.models.detection import FasterRCNN_ResNet50_FPN_V2_Weights,\
+    RetinaNet_ResNet50_FPN_Weights
+
+
+class DetectionModelsWeights:
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/configs.py]: Maybe use Enum?

File: pytorch_object_detection/detection_models/configs.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+    RESNET = FasterRCNN_ResNet50_FPN_V2_Weights.DEFAULT
+    RETINANET = RetinaNet_ResNet50_FPN_Weights.DEFAULT
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/configs.py]: New line

File: pytorch_object_detection/detection_models/detection_models.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        super().__init__()
+
+
+if __name__ == '__main__':
+
+    path = "../test_data/cat_duck_dog.jpg"
+
+    model = ResNetDetection()
+    labels, boxes, scores = model.run(path, score_threshold=0.5)
+
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/detection_models.py]: Move it to separate run script

**Comment by AlexShashko** [comment]
Add docstrings

File: pytorch_object_detection/detection_models/__init__.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    @staticmethod
+    def visualize_result(imgs, boxes, labels):
+        for ind, img in enumerate(imgs):
+            box = torch.stack(boxes[ind], dim=0)
+            drawed_boxes = draw_bounding_boxes(img, boxes=box,
+                                               labels=labels[ind],
+                                               width=2)
+            im = to_pil_image(drawed_boxes.detach())
+            im.show(title=f'Image №{ind}')
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]:
```suggestion
            im.show(title=f'Image #{ind}')
```

File: pytorch_object_detection/detection_models/__init__.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        for output in outputs:
+            labels, boxes, scores = [], [], []
+            for ind, score in enumerate(output['scores']):
+                if score >= score_threshold:
+                    scores.append(score)
+                    labels.append(self.model_classes[output["labels"][ind]])
+                    boxes.append(output['boxes'][ind])
+            output_scores.append(scores)
+            output_labels.append(labels)
+            output_boxes.append(boxes)
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Move it also to private method

**Review by AlexShashko — CHANGES_REQUESTED** [review]

File: pytorch_object_detection/detection_models/detection_models.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        self.model = retinanet_resnet50_fpn(
+            weights=self.weights,
+
+        )
+        super().__init__()
+
+
+if __name__ == '__main__':
+
+    path = "../test_data/cat_duck_dog.jpg"
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/detection_models.py]: Use argparse

**Review by AlexShashko — CHANGES_REQUESTED** [review]

File: pytorch_object_detection/detection_models/__init__.py (hunk 3 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: done

File: pytorch_object_detection/detection_models/__init__.py (hunk 4 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: done

File: pytorch_object_detection/detection_models/__init__.py (hunk 5 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: done

File: pytorch_object_detection/detection_models/configs.py (hunk 7 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/configs.py]: done

File: pytorch_object_detection/detection_models/configs.py (hunk 8 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/configs.py]: done

File: pytorch_object_detection/detection_models/detection_models.py (hunk 9 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/detection_models.py]: done

File: pytorch_object_detection/detection_models/__init__.py (hunk 10 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: done

File: pytorch_object_detection/detection_models/__init__.py (hunk 11 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: done

File: pytorch_object_detection/detection_models/__init__.py (hunk 2 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: As only 'model' and 'weights' need to be defined in child class init method  and call parent init to define preprocess_pipeline and model_classes so I added it as hint for defining child init, or it is better only leave  
model: torchvision.models.detection
weights: WeightsEnum 
What do you think?

File: pytorch_object_detection/README.md (hunk 1 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/README.md]: done?

File: pytorch_object_detection/detection_models/configs.py (hunk 7 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/configs.py]: When use Enum I can not get meta of classess of weights, so I removed it

File: pytorch_object_detection/detection_models/detection_models.py (hunk 12 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/detection_models.py]: done

File: pytorch_object_detection/detection_models/__init__.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+from typing import List
+
+
+class DetectionPipeline(ABC):
+    """
+    Class defining basic functionality to run pretrained object detection model on image or batch of images.
+    In __init__ func of inherit class it is needed to:
+        1. define self.weights of pretrained model
+        2. define self.model with these weights
+        3. call init of this class
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Move/copy to __ init __ docstring

File: pytorch_object_detection/detection_models/__init__.py (hunk 2 above)
Comment (AlexShashko) [reply pytorch_object_detection/detection_models/__init__.py:25]: Move it

File: pytorch_object_detection/detection_models/__init__.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+    def __init__(self):
+        self.model.eval()
+        self.model_classes = self.weights.meta["categories"]
+        self.preprocess_pipeline = self.weights.transforms()
+
+    def _image_preprocess(self, image_paths: List[str] | str) -> (List[Tensor], List[Tensor]):
+        """
+        Read and preprocesses images for the model
+        :param image_paths: path's to images or image
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]:
```suggestion
        Read and preprocesses images for the model
        
        :param image_paths: path's to images or image
```

File: pytorch_object_detection/detection_models/__init__.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+        raw_images = [read_image(path) for path in image_paths]
+        batch = [self.preprocess_pipeline(img) for img in raw_images]
+
+        return raw_images, batch
+
+    @staticmethod
+    def draw_boxes(imgs: List[Tensor], boxes: List[list], labels: List[List[str]]) -> list:
+        """
+        Visualize detected objects - image with predicted boxes and labels
+        :param imgs: image on which objects wer detected
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]:
```suggestion
        Visualize detected objects - image with predicted boxes and labels
        
        :param imgs: image on which objects wer detected
```

File: pytorch_object_detection/detection_models/configs.py (hunk 7 above)
Comment (AlexShashko) [reply pytorch_object_detection/detection_models/configs.py:5]: Do Enum, if any issues - log here

File: pytorch_object_detection/detection_models/models.py [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+class FasterRCNNDetection(DetectionPipeline):
+    """
+    Pretrained Faster R-CNN for object detection.
+    """
+    def __init__(self):
+        self.weights = DetectionModelsWeights.FASTER_RCNN
+        self.model = fasterrcnn_resnet50_fpn_v2(
+            weights=self.weights
+        )
+        super().__init__()
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/models.py]: Modify __ init __

File: pytorch_object_detection/detection_models/models.py [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+class RetinaNetDetection(DetectionPipeline):
+    """
+    Pretrained RetinaNet for object detection.
+    """
+    def __init__(self):
+        self.weights = DetectionModelsWeights.RETINANET
+        self.model = retinanet_resnet50_fpn_v2(
+            weights=self.weights
+        )
+        super().__init__()
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/models.py]: Same

File: pytorch_object_detection/detection_models/run.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
+import argparse
+import os
+import json
+from pytorch_object_detection.detection_models.models import FasterRCNNDetection, RetinaNetDetection
+
+
+def process_console_run(console_input: argparse.Namespace) -> None:
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]:
```suggestion
def process_console_run(console_input: argparse.Namespace):
```

File: pytorch_object_detection/detection_models/__init__.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+                    labels.append(self.model_classes[output["labels"][ind]])
+                    boxes.append(output['boxes'][ind].tolist())
+
+            filtered_output['scores'].append(scores)
+            filtered_output['labels'].append(labels)
+            filtered_output['boxes'].append(boxes)
+
+        return filtered_output
+
+    def run(self, image_paths: List[str] | str, score_threshold: float = 0.9, to_visualize: bool = True) -> (
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Check (dict, list)

File: pytorch_object_detection/detection_models/run.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    # Select model
+    match console_input.model_name:
+        case 'Faster R-CNN':
+            model = FasterRCNNDetection()
+        case 'RetinaNet':
+            model = RetinaNetDetection()
+        case _:
+            print('This model is not implemented')
+            return
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]: Make Enum with model names and develop from_str classmethod inside

File: pytorch_object_detection/detection_models/run.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+    # Get detection predictions
+    image_names = os.listdir(path=console_input.source_folder)
+    paths_to_images = [console_input.source_folder + '/' + name for name in image_names]
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]: use os.path.join()

File: pytorch_object_detection/detection_models/run.py [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+    predictions, images = model.run(
+        paths_to_images,
+        score_threshold=console_input.score_threshold
+    )
+
+    # Save predictions in json file
+    if not os.path.exists(console_input.save_folder_path):
+        os.mkdir(console_input.save_folder_path)
+    out_json = open(console_input.save_folder_path + '/boxes_labels_scores.json', "w")
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]: same

File: pytorch_object_detection/detection_models/run.py [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    json.dump(predictions, out_json)
+
+    # Save images with drawn boxes
+    dir_for_images = console_input.save_folder_path + '/drawn_boxes/'
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]: same

File: pytorch_object_detection/detection_models/run.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+
+    # Get detection predictions
+    image_names = os.listdir(path=console_input.source_folder)
+    paths_to_images = [console_input.source_folder + '/' + name for name in image_names]
+    predictions, images = model.run(
+        paths_to_images,
+        score_threshold=console_input.score_threshold
+    )
+
+    # Save predictions in json file
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]: Add to separate function

File: pytorch_object_detection/detection_models/run.py (hunk 18 above)
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]: Replace consol_input by kwargs

File: pytorch_object_detection/README.md (hunk 1 above)
Comment (AlexShashko) [reply pytorch_object_detection/README.md]: Information about you

Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/README.md]: done

File: pytorch_object_detection/detection_models/__init__.py (hunk 2 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: done

File: pytorch_object_detection/detection_models/configs.py (hunk 7 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/configs.py]: done

File: pytorch_object_detection/detection_models/__init__.py (hunk 13 above)
Comment (OvsiienkoVladyslava) [reply pytorch_object_detection/detection_models/__init__.py]: done

File: README.md [hunk 25]
```diff
@@ -1,2 +1,13 @@
-# computer_vision
-Test and try CV models
+# Computer Vision Pet Projects
+The goal of this repository is to solve some Computer Vision tasks using pre-trained models or train new.
```
Comment (AlexShashko) [inline README.md]:
```suggestion
The goal of this repository is to solve some Computer Vision tasks using pre-trained models or train new ones.
```

File: flake8.cfg [hunk 26]
```diff
@@ -0,0 +1,3 @@
+[flake8]
+max-line-length = 120
+ignore = E203, W503
```
Comment (AlexShashko) [inline flake8.cfg]: Add new line

File: pytorch_object_detection/detection_models/__init__.py [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+        self.preprocess_pipeline = self.weights.transforms()
+
+    def _image_preprocess(self, image_paths: List[str] | str) -> (List[Tensor], List[Tensor]):
+        """
+        Read and preprocesses images for the model
+
+        :param image_paths: path's to images or image
+        :return: read images and preprocessed for detection model
+        """
+        if type(image_paths) is str:
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Maybe better isinstance(obj, type)?

File: pytorch_object_detection/detection_models/__init__.py [hunk 28]
```diff
@@ -0,0 +1,150 @@
...
+        Run predictions of the model.
+
+        :param to_visualize: to visualize results or not
+        :param image_paths: path's to images or image on which detect objects
+        :param score_threshold: threshold of confidence to filter results (leave if <= threshold)
+        :return: predicted labels, boxes and confidence scores
+        """
+        # Read and preprocess images
+        raw_images, batch = self._image_preprocess(image_paths)
+        image_names = [path.split("\\")[-1] for path in image_paths]
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]: Replace "\\\\" with os.sep

File: pytorch_object_detection/detection_models/__init__.py (hunk 3 above)
Comment (AlexShashko) [reply pytorch_object_detection/detection_models/__init__.py]: Still not abstract.
Add load_model(...) abstract method to set up self.model and self.weights in inherited classes

File: pytorch_object_detection/detection_models/__init__.py [hunk 29]
```diff
@@ -0,0 +1,150 @@
...
+
+            filtered_output["scores"].append(scores)
+            filtered_output["labels"].append(labels)
+            filtered_output["boxes"].append(boxes)
+
+        return filtered_output
+
+    def run(
+        self, image_paths: List[str] | str, score_threshold: float = 0.9, to_visualize: bool = True
+    ) -> (dict, list):
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]:
```suggestion
    ) -> Tuple[dict, list]:
```

File: pytorch_object_detection/detection_models/__init__.py [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+        Run predictions of the model.
+
+        :param to_visualize: to visualize results or not
+        :param image_paths: path's to images or image on which detect objects
+        :param score_threshold: threshold of confidence to filter results (leave if <= threshold)
+        :return: predicted labels, boxes and confidence scores
+        """
+        # Read and preprocess images
+        raw_images, batch = self._image_preprocess(image_paths)
+        image_names = [os.path.split(path)[1] for path in image_paths]
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/__init__.py]:
```suggestion
        image_names = [os.path.basename(path) for path in image_paths]
```

File: pytorch_object_detection/detection_models/configs.py [hunk 31]
```diff
@@ -0,0 +1,24 @@
+from enum import Enum
+
+from pytorch_object_detection.detection_models.models import (
+    FasterRCNNDetection,
+    RetinaNetDetection,
+)
+
+
+class DetectionModelsNames(Enum):
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/configs.py:9]: Add class description

File: pytorch_object_detection/detection_models/models.py [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+    FasterRCNN_ResNet50_FPN_V2_Weights,
+    RetinaNet_ResNet50_FPN_V2_Weights,
+    fasterrcnn_resnet50_fpn_v2,
+    retinanet_resnet50_fpn_v2,
+)
+
+from pytorch_object_detection.detection_models import DetectionPipeline
+
+
+class DetectionModelsWeights(Enum):
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/models.py:15]: Add class description

File: pytorch_object_detection/detection_models/run.py [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+    predictions, images = model.run(image_paths=paths_to_images, score_threshold=score_threshold, to_visualize=False)
+
+    # Save predictions in json file and images with drawn boxes
+    model.save_results(save_folder_path=save_folder_path, predictions=predictions, images=images)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser("Run inference of selected detection model")
+    parser.add_argument(
+        "--source-folder", type=str, default=".\\test_data", help='path to folder with images, e.g. ".\\test_data" '
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]:
```suggestion
        "--source-folder", type=str, default="./test_data", help='path to folder with images, e.g. "./test_data" '
```

File: pytorch_object_detection/detection_models/run.py [hunk 34]
```diff
@@ -0,0 +1,46 @@
...
+    parser.add_argument(
+        "--model-name", type=str, default="Faster RCNN", help="model name to use, available: Faster RCNN, RetinaNet"
+    )
+    parser.add_argument(
+        "--score-threshold", type=float, default=0.9, help="min confidence threshold of prediction, e.g 0.8"
+    )
+    parser.add_argument(
+        "--save-folder-path",
+        type=str,
+        default=".\\test_results",
```
Comment (AlexShashko) [inline pytorch_object_detection/detection_models/run.py]:
```suggestion
        default="./test_results",
```

**Review by AlexShashko — APPROVED** [review]

**Pull request merged**